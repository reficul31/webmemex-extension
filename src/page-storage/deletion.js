import db from 'src/pouchdb'
import { findVisits } from 'src/search/find-visits'
import { getEquivalentPages } from 'src/search/find-pages'

/**
* Deletes a visit and its page.
*
* @param {string} visitId - visit ID of the visit to be deleted
*/
export async function deleteVisitAndPage({visitId}) {
    // Delete the visit object
    const visit = await db.get(visitId)
    const pageId = visit.page._id
    await db.remove(visit)

    // If this was the only visit linking to the page, also remove the page.
    // (a simple choice for now; this behaviour may be changed in the future)
    await deletePageIfOrphaned({pageId})
}

/**
* Delete a page if it is orphaned
*
* @param {string} pageId - pageId of the page to be checked
*/
async function deletePageIfOrphaned({pageId}) {
    // Because of deduplication, different page objects may redirect to this
    // one, or this one may redirect to others. So we check for visits either
    // referring to this page object or to any equivalent ones.
    const pagesResult = await getEquivalentPages({pageId})
    const visitsResult = await findVisits({pagesResult})
    // If there are no visits to any of them, delete all these page objects.
    // (otherwise, we leave them; it does not seem worth the effort to smartly
    // prune some orphans among them already)
    if (visitsResult.rows.length === 0) {
        await db.bulkDocs(pagesResult.rows.map(
            row => ({
                _id: row.id,
                _rev: row.value.rev,
                _deleted: true,
            })
        ))
    }
}
