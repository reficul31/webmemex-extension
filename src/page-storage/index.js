import docuri from 'docuri'
import randomString from 'src/util/random-string'

export const pageKeyPrefix = 'page/'

export const convertPageDocId = docuri.route(`${pageKeyPrefix}:timestamp/:nonce`)

/**
* Generate a page doc id
*
* @param {Number} timestamp - timestamp of the page
* @param {string} nonce - random string of 10 digit number
*
* @returns {string} - doc id of the page as a string
*/
export function generatePageDocId({timestamp, nonce} = {}) {
    const date = timestamp ? new Date(timestamp) : new Date()
    return convertPageDocId({
        timestamp: date.getTime(),
        nonce: nonce || randomString(),
    })
}
