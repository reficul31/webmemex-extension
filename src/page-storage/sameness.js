import get from 'lodash/fp/get'
import DiffMatchPatch from 'diff-match-patch'

export const Sameness = {
    EXACTLY: 5,    // Perfect match, containing exactly the same data.
    OSTENSIBLY: 4, // Semantically the same, differing only in e.g. whitespace.
    MOSTLY: 3,     // Only minor visible edits, e.g. some typo corrections.
    PARTLY: 2,     // Some content changed, e.g. parts were added or (re)moved.
    HARDLY: 1,     // Whole page seems newly created, but with similar purpose.
    UNRELATED: 0,  // Completely different topic, e.g. domain was transferred.
    UNKNOWN: NaN,  // Not enough information to make a comparison.
}

/**
* Get a rough similarity metric between strings (a number between 0 and 1).
*
* @param {string} text1 -first text to be compared
* @param {string} text2 -second text to be compared
*
* @returns {Number} - quntifies the similarity of the two texts
*/
function stringSimilarity(text1, text2) {
    const dmp = new DiffMatchPatch()
    dmp.Diff_Timeout = 0.1
    const diff = dmp.diff_main(text1, text2)
    dmp.diff_cleanupSemantic(diff)
    const distance = dmp.diff_levenshtein(diff)
    const maxDistance = Math.max(text1.length, text2.length)
    const similarity = 1 - distance / maxDistance
    return similarity
}

/**
* Tell how similar two strings are in a qualitative way.
*
* @param {string} text1 -first text to be compared
* @param {string} text2 -second text to be compared
*
* @returns {Number} - quntifies the similarity of the two texts
*/
function textSameness(text1, text2) {
    if (!text1 || !text2) { return Sameness.UNKNOWN }

    if (text1 === text2) { return Sameness.EXACTLY }

    const normaliseWhitespace = s => s.trim().replace(/s+/g, ' ')
    if (normaliseWhitespace(text1) === normaliseWhitespace(text2)) { return Sameness.OSTENSIBLY }

    const textSimilarity = stringSimilarity(text1, text2)

    if (textSimilarity > 0.9) { return Sameness.MOSTLY }

    if (textSimilarity > 0.7) { return Sameness.PARTLY }

    if (textSimilarity > 0.5) { return Sameness.HARDLY }

    return Sameness.UNRELATED
}

/**
* Tell to which degree two versions of a page can be considered to be the same.
*
* @param {string} page1 -first Page to be compared
* @param {string} page2 -second Page to be compared
*
* @returns {Number} - quntifies the similarity of the two Pages
*/
export default function determinePageSameness(page1, page2) {
    const pages = [page1, page2]

    // We start from the maximum score, and lower it when we discover differences.
    let score = Sameness.EXACTLY

    // First a quick check for exact equality of the stored 'freeze-dried' pages.
    const hashes = pages.map(get(['_attachments', 'frozen-page.html', 'digest']))
    if (hashes.every(hash => (hash === hashes[0]))) {
        if (hashes[0] !== undefined) {
            // If the stored pages are exactly the same, no need to compare their contents.
            return Sameness.EXACTLY
        } else {
            // If neither page was freeze-dried, we ignore just this test and compare the content.
        }
    } else {
        // The stored pages differ to some unknown extent. We go on to compare the content, but
        // we lower the maximum possible score so we cannot return EXACTLY.
        score = Sameness.OSTENSIBLY
    }

    // Compare extracted content, and return the lowest score among them.
    const fieldsToCompare = {
        'content.title': textSameness,
        'content.fullText': textSameness,
    }
    for (let field in fieldsToCompare) {
        const compare = fieldsToCompare[field]
        const fieldSameness = compare(...pages.map(get(field)))
        score = Math.min(score, fieldSameness)
    }

    return score
}
