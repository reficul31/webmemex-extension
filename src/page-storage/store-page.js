import db from 'src/pouchdb'
import { findPagesByUrl } from 'src/search/find-pages'
import analysePage from 'src/page-analysis/background'
import tryDedupePage from './deduplication'
import { generatePageDocId } from '.'

/**
* Try see if we can tell in advance that this page will be one we already know.
* Returns the matching page doc, if any.
*
* @param {Number} tabId - tab ID of the active page
* @param {string} url - URL of the page to be reidentified
* @param {Array} samePageCandidates - Array of Pages
*/
async function tryReidentifyPage({tabId, url, samePageCandidates}) {
    // TODO check ETag or other proof of equality.

    // No match found. We will have to make a new page.
    return undefined
}

/**
* Create a page stub in the database and return a page
*
* @param {string} url - URL of the page stub to be created
*
* @returns {Page} - Page instance of the current page
*/
async function createPageStub({url}) {
    const pageId = generatePageDocId()
    const page = {
        _id: pageId,
        url,
    }
    await db.put(page)
    return page
}

/**
* Analyse the page and Deduplicate it
*
* @param {Number} tabId - tab ID of the active page
* @param {Page} page - Page instance to be analysed and deduplicated
* @param {Array} samePageCandidates - Array of Pages containing similar pages
*
* @returns {Page} - final Page instance after deduplication and analysis
*/
async function analyseAndTryDedupePage({tabId, page, samePageCandidates}) {
    // Add info to the page doc by analysing the document in the tab.
    const {page: analysedPage} = await analysePage({page, tabId})

    // Knowing more about the page now, try find it again in our memory.
    const {page: finalPage} = await tryDedupePage({page: analysedPage, samePageCandidates})

    // Return the resulting page (likely still the same as analysedPage)
    return {page: finalPage}
}

/**
* Tries to reidentify the page in the tab, or creates a new page doc for it.
* Either way, it returns the page doc. If a new page was created, it starts
* page analysis and then deduplication, and also returns a promise
* (finalPagePromise) of the page doc resulting from these steps.
*
* @param {Number} tabId - tab ID of the active page
* @param {string} url - url of the page to be stored
*
* @returns {Page} - Page instance of the page
* @returns {finalPagePomise} - Promise that resolves to an analysed and deduplicated page
*/
export async function reidentifyOrStorePage({tabId, url}) {
    // Find pages we know that had the same URL.
    const samePageCandidates = (await findPagesByUrl({url})).rows.map(row => row.doc)

    // Check if we can already tell in advance that this is the same page.
    const reusablePage = await tryReidentifyPage({tabId, url, samePageCandidates})

    if (reusablePage) {
        // Equality is known (or assumed) in advance. Reuse the old page as is.
        return {page: reusablePage}
    } else {
        // Create a new page doc in the database.
        const page = await createPageStub({url})

        // Start analysis and (possibly) deduplication, but do not wait for it.
        const finalPagePromise = analyseAndTryDedupePage({
            tabId,
            page,
            samePageCandidates,
        })

        // Return the page stub, and a promise of the analysed & deduped page.
        return {page, finalPagePromise}
    }
}
