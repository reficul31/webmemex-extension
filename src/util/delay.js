/**
* Creates a delay of the timeperiod specified
*
* @param {Number} t - timeperiod for the delay
*
* @returns {Promise} - resolves after a timeout
*/
export default function delay(t) {
    return new Promise(
        resolve => window.setTimeout(resolve, t)
    )
}
