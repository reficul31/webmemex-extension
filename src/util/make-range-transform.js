import clamp from 'lodash/fp/clamp'
import compose from 'lodash/fp/compose'

/**
* Create an affine transformation to map numbers from one range to another.
*
* @param {Array} domain - Domain range of values for input
* @param {Array} range - Range of values for mapping
* @param {boolean} clampOutput - whether the values above or below the range be capped or not
*
* @returns {Function} - range transform function
*/
export function makeRangeTransform({domain: [minX, maxX], range: [minY, maxY], clampOutput = false}) {
    const scale = (maxY - minY) / (maxX - minX)
    const bias = minY - minX * scale
    const transform = x => bias + x * scale
    return clampOutput ? compose(clamp(minY, maxY), transform) : transform
}

/**
* Create a transformation that passes the input through a monotonic nonlinearity
*
* @param {Array} domain - Domain range of values for input
* @param {Array} range - Range of values for mapping
* @param {Function} nonlinearity - non linearity function
* @param {boolean} clampOutput - whether the values above or below the range be capped or not
*
* returns {function} - non linear transform function
*/
export function makeNonlinearTransform({domain, range, nonlinearity, clampOutput = false}) {
    return function nonLinearTransform(input) {
        const toOutputRange = makeRangeTransform({
            domain: domain.map(nonlinearity),
            range,
            clampOutput,
        })
        return toOutputRange(nonlinearity(input))
    }
}
