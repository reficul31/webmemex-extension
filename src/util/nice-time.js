/**
* Returns the hour as a string
*
* @param {Date} date - date to be processed
*
* @returns {string} - hour of the Date as a string
*/
function hourString(date) {
    // return date.toLocaleTimeString([], {hour: 'numeric', minute: '2-digit'})
    return `${('0' + date.getHours()).slice(-2)}:${('0' + date.getMinutes()).slice(-2)}`
}

/**
* Returns the day as a string
*
* @param {Date} date - date to be processed
*
* @returns {string} - day of the Date as a string
*/
function dayString(date) {
    const days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat']
    return days[date.getDay()]
}

/**
* Returns the month as a string
*
* @param {Date} date - date to be processed
*
* @returns {string} - month of the Date as a string
*/
function monthString(date) {
    const months = ['Jan', 'Feb', 'March', 'Apr', 'May', 'June', 'July', 'Aug', 'Sept', 'Oct', 'Nov', 'Dec']
    return months[date.getMonth()]
}

/**
* Used to format the time format to get a readable string
*
* @param {Date} date - date to be processed
* @param {Date} now - date from where the difference is to be measured
*
* @returns {string} - a human redable formatted string for the time of the Date instance from now
*/
export default function niceTime(date, {now = undefined} = {}) {
    const then = new Date(date)
    now = now || new Date()
    const secondsAgo = (now - then) / 1000
    if (secondsAgo < -60 * 60 * 24) {
        return 'in the future?!'
    }
    if (secondsAgo < -60) {
        return 'soon?!'
    }
    if (secondsAgo < 90) { return 'now' }
    if (secondsAgo < 60 * 10) { return `${Math.round(secondsAgo / 60)} minutes ago` }
    if (secondsAgo < 60 * 60) { return `${Math.round(secondsAgo / 60 / 5) * 5} minutes ago` }
    if (secondsAgo < 60 * 60 * 24
        && (now.getDay() === then.getDay() || secondsAgo < 60 * 60 * 6)) { return hourString(then) }
    if (secondsAgo < 60 * 60 * 24) { return `Yesterday ${hourString(then)}` }
    if (secondsAgo < 60 * 60 * 24 * 3) { return `${dayString(then)} ${hourString(then)}` }
    if (then.getYear() === now.getYear()) { return `${then.getDate()} ${monthString(then)}` }
    return `${then.getDate()} ${monthString(then)} ${then.getFullYear()}`
}

/**
* Used to format the date format to get a readable string
*
* @param {Date} date - date to be processed
* @param {Date} now - date from where the difference is to be measured
*
* @returns {string} - a human redable formatted string for the date of the Date instance from now
*/
export function niceDate(date, {now = new Date()} = {}) {
    const then = new Date(date)
    if (then.getYear() !== now.getYear()) {
        return `${then.getDate()} ${monthString(then)} ${then.getFullYear()}`
    }
    if (then.getMonth() === now.getMonth()) {
        const daysAgo = now.getDate() - then.getDate()
        if (daysAgo === 0) return `Today`
        if (daysAgo === 1) return `Yesterday`
    }
    return `${dayString(then)} ${then.getDate()} ${monthString(then)}`
}
