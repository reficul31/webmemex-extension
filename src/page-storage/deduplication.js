// See for explanation: https://github.com/WebMemex/webmemex-extension/issues/22#issuecomment-282329412

import get from 'lodash/fp/get'
import maxBy from 'lodash/fp/maxBy'

import db from 'src/pouchdb'
import updateDoc from 'src/util/pouchdb-update-doc'
import { getTimestamp } from 'src/activity-logger'
import determinePageSameness, { Sameness } from './sameness'
/**
* Quantifies the content matching of a page
*
* @param {Number} sameness - the degree of sameness of two pages
*
* @returns {string} - the degree of sameness as a string
*/
function samenessLinkType({sameness}) {
    const types = {
        [Sameness.EXACTLY]: 'sameAs',
        [Sameness.OSTENSIBLY]: 'updatedBy',
        [Sameness.MOSTLY]: 'updatedBy',
        [Sameness.PARTLY]: 'updatedBy',
        [Sameness.HARDLY]: 'replacedBy',
        [Sameness.UNRELATED]: 'replacedBy',
        [Sameness.UNKNOWN]: undefined,
    }
    return types[sameness]
}

/**
* Remove the stored content of the page. Eg.Screenshots, page content, etc.
*
* @param {Page} page - Page instance of the page
* @param {string} pageId - page ID of the page
*/
async function forgetPageContents({page, pageId = page._id}) {
    // Remove stored content, screenshots, etcetera.
    await updateDoc(db, pageId, doc => ({
        ...doc,
        content: undefined,
        _attachments: {},
    }))
}

/**
* Adds a link to the document
*
* @param {string} targetId -
* @param {string} sourceId -
* @param {string} linkType -
*/
async function addLink({targetId, sourceId, linkType}) {
    await updateDoc(db, sourceId, doc => ({
        ...doc,
        [linkType]: {_id: targetId},
    }))
}

/**
* Replace the page with a redirected page.
*
* @param {Page} oldPage -
* @param {Page} newPage -
* @param {Number} sameness -
*/
async function replaceWithRedirect({oldPage, newPage, sameness}) {
    await forgetPageContents({page: oldPage})
    // Add a reference that indicates where to find usable content instead.
    await addLink({
        targetId: newPage._id,
        sourceId: oldPage._id,
        linkType: 'seeInstead',
    })
    // Also express the type of relation (degree of sameness) between the pages.
    await addLink({
        targetId: newPage._id,
        sourceId: oldPage._id,
        linkType: samenessLinkType({sameness}),
    })
}

/**
* Tells whether the record quality of one page is better than that of the other.
*
* @param {Page} page -
* @param {Page} comparisonPage -
*
* @returns {boolean} -
*/
function hasHigherFidelity(page, comparisonPage) {
    const hasFrozenPage = page =>
        (get(['_attachments', 'frozen-page.html'])(page) !== undefined)
    if (hasFrozenPage(page) && !hasFrozenPage(comparisonPage)) {
        // page was successfully freeze-dried, while comparisonPage was not.
        return true
    }
    return false
}

/**
* Try and deduplicate a given page with its same page candidates.
*
* @param {Page} page -
* @param {Array} samePageCandidates -
*
* @returns {Page} -
*/
export default async function tryDedupePage({
    page,
    samePageCandidates,
}) {
    if (samePageCandidates.length === 0) {
        return {page}
    }

    // For simplicity, only try the most recent candidate.
    const candidatePage = maxBy(doc => getTimestamp(doc))(samePageCandidates)

    // Measure how similar they are.
    const sameness = determinePageSameness(page, candidatePage)

    // Choose the action to take.
    if (
        sameness >= Sameness.EXACTLY
        && !candidatePage.protected
        // Even if they seem to represent the same content, ensure we don't delete a better copy.
        && !hasHigherFidelity(candidatePage, page)
    ) {
        // Forget the old page's contents. Replace them with a link to the new
        // page.
        await replaceWithRedirect({
            oldPage: candidatePage,
            newPage: page,
            sameness,
        })
    } else {
        // Do nothing, leave as duplicate. We do create a link for information
        // if they are similar enough.
        if (sameness >= Sameness.PARTLY) {
            await addLink({
                sourceId: candidatePage._id,
                targetId: page._id,
                linkType: samenessLinkType({sameness}),
            })
        }
    }
    // Same page is returned.
    return {page}
}
