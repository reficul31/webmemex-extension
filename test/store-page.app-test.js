/* eslint-env jest */
/* eslint import/namespace: "off" */

import { reidentifyOrStorePage } from 'src/page-storage/store-page'
import { dataURLToBlob } from 'blob-util'
import * as webextensionRPC from 'src/util/webextensionRPC'
import db from 'src/pouchdb'

const imageDataUri = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAAAAAA6fptVAAAACklEQVR4nGNiAAAABgADNjd8qAAAAABJRU5ErkJggg=='

jest.mock('src/util/tab-events')

browser.tabs = {
    get: jest.fn().mockReturnValue({
        windowId: 1,
        favIconUrl: 'https://example.com/icon.ico',
    }),
    captureVisibleTab: jest.fn().mockReturnValue(imageDataUri),
}

const removeDynamicFields = (page) => {
    delete page.page._id
    delete page.page._rev
}

beforeEach(() => {
    const freezeDry = jest.fn().mockReturnValue(
        new Promise((resolve, reject) => {
            resolve('<html></html>')
        })
    )
    const extractPageContent = jest.fn().mockReturnValue(
        new Promise((resolve, reject) => {
            resolve('page content')
        })
    )
    webextensionRPC.remoteFunction = jest.fn().mockReturnValueOnce(extractPageContent).mockReturnValueOnce(freezeDry)
})

afterAll(async () => {
    await db.destroy()
})

describe('App tests', () => {
    const tabId = 1
    const url = 'https://example.com/page'
    let favIcon

    beforeAll(async () => {
        favIcon = await dataURLToBlob(imageDataUri)
    })

    test('should return a page stub and finalPage when no similar page is in the database', async () => {
        fetch.mockResponse(favIcon)
        let expectedId = expect.stringMatching('^page/[0-9]{13}/[0-9]{10}$')
        const { page, finalPagePromise } = await reidentifyOrStorePage({tabId, url})
        const finalPage = await finalPagePromise
        expect(page.url).toBe('https://example.com/page')
        expect(page._id).toEqual(expectedId)
        removeDynamicFields(finalPage)
        expect(finalPage).toMatchSnapshot()
    })

    test('should return a page stub and finalPage when a similar page is in the database', async () => {
        fetch.mockResponse(favIcon)
        let expectedId = expect.stringMatching('^page/[0-9]{13}/[0-9]{10}$')
        const { page, finalPagePromise } = await reidentifyOrStorePage({tabId, url})
        const finalPage = await finalPagePromise
        expect(page.url).toBe('https://example.com/page')
        expect(page._id).toEqual(expectedId)
        removeDynamicFields(finalPage)
        expect(finalPage).toMatchSnapshot()
    })
})
