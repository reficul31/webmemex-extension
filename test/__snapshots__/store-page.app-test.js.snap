// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests should return a page stub and finalPage when a similar page is in the database 1`] = `
Object {
  "page": Object {
    "_attachments": Object {
      "favIcon": Object {
        "content_type": "image/png",
        "digest": "md5-7cVkAmmX4suBnAFSJ4A2Wg==",
        "length": 67,
        "revpos": 4,
        "stub": true,
      },
      "frozen-page.html": Object {
        "content_type": "text/html;charset=utf-8",
        "digest": "md5-yDMBQlsq0dSWRzpf89nsyg==",
        "length": 13,
        "revpos": 3,
        "stub": true,
      },
      "screenshot": Object {
        "content_type": "image/png",
        "digest": "md5-7cVkAmmX4suBnAFSJ4A2Wg==",
        "length": 67,
        "revpos": 5,
        "stub": true,
      },
    },
    "content": "page content",
    "title": "",
    "url": "https://example.com/page",
  },
}
`;

exports[`App tests should return a page stub and finalPage when no similar page is in the database 1`] = `
Object {
  "page": Object {
    "_attachments": Object {
      "favIcon": Object {
        "content_type": "image/png",
        "digest": "md5-7cVkAmmX4suBnAFSJ4A2Wg==",
        "length": 67,
        "revpos": 4,
        "stub": true,
      },
      "frozen-page.html": Object {
        "content_type": "text/html;charset=utf-8",
        "digest": "md5-yDMBQlsq0dSWRzpf89nsyg==",
        "length": 13,
        "revpos": 3,
        "stub": true,
      },
      "screenshot": Object {
        "content_type": "image/png",
        "digest": "md5-7cVkAmmX4suBnAFSJ4A2Wg==",
        "length": 67,
        "revpos": 5,
        "stub": true,
      },
    },
    "content": "page content",
    "title": "",
    "url": "https://example.com/page",
  },
}
`;
